### PR TITLE
[TAS-1863] ✨ Add character count limit for author's message

### PR DESCRIPTION
**Pull request by AuroraHuang22**
截圖 2024-10-08 晚上7 34 43
截圖 2024-10-08 晚上7 34 51

### CODE REVIEW

File: pages/nft-book-store/send/[classId].vue [hunk 1]
```diff
@@ -199,9 +208,17 @@ const orderInfo = ref<any>({})
...
 
 const nftClassName = computed(() => nftStore.getClassMetadataById(classId.value as string)?.name)
+const messageCharCount = computed(() => {
+  let count = 0
+  for (let i = 0; i < memo.value.length; i++) {
+    count += isFullWidth(memo.value[i]) ? 2 : 1
+  }
+  return count
+})
+const isLimitReached = computed(() => { return messageCharCount.value > AUTHOR_MESSAGE_LIMIT })
```
Comment (nwingt) [inline pages/nft-book-store/send/[classId].vue:221]: I think these 2 functions can be a [composable](https://vuejs.org/guide/reusability/composables#composables)

**Review by nwingt — CHANGES_REQUESTED** [review]

File: pages/nft-book-store/collection/send/[collectionId].vue [hunk 2]
```diff
@@ -89,7 +89,14 @@
...
       <UFormGroup label="Enter Author's Message" hint="optional">
-        <UTextarea v-model="memo" placeholder="default memo" />
+        <UTextarea
+          v-model="memo"
+          placeholder="default memo"
+          :color="isLimitReached? 'red' : 'primary'"
+        />
+        <p class="text-gray-500 text-[12px]">
+          {{ `${messageCharCount} / ${AUTHOR_MESSAGE_LIMIT}` }}
+        </p>
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply pages/nft-book-store/collection/send/[collectionId].vue]: [🎨 Update components error & hint UI using default props](https://github.com/likecoin/nft-book-press/pull/212/commits/76395735d2608ad913a97667d5a2fa7aee65155c)

File: pages/nft-book-store/send/[classId].vue [hunk 3]
```diff
@@ -376,4 +393,8 @@ async function onSendNFTStart () {
   }
 }
 
+function isFullWidth (char: any) {
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply pages/nft-book-store/send/[classId].vue]: [🎨 Update components error & hint UI using default props](https://github.com/likecoin/nft-book-press/pull/212/commits/76395735d2608ad913a97667d5a2fa7aee65155c)

File: pages/nft-book-store/send/[classId].vue (hunk 1 above)
Comment (AuroraHuang22) [reply pages/nft-book-store/send/[classId].vue]: [♻️ Refactor char count computed to composables](https://github.com/likecoin/nft-book-press/pull/212/commits/07a063f483d706b134ffd27e42dfe599c3b335b4)

File: composables/useMessageCharCount.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+interface UseMessageCharCountReturn {
+  messageCharCount: ComputedRef<number>;
+  isLimitReached: ComputedRef<boolean>;
+}
+
+function isFullWidthChar (char: string): boolean {
+  return char.charCodeAt(0) > 255
+}
+
+export function useMessageCharCount (memo: Ref<string>, AUTHOR_MESSAGE_LIMIT: number): UseMessageCharCountReturn {
```
Comment (nwingt) [inline composables/useMessageCharCount.ts:12]:
```suggestion
export function useMessageCharCount (memo: Ref<string>, limit: number): UseMessageCharCountReturn {
```

File: composables/useMessageCharCount.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  const messageCharCount = computed((): number => {
+    let count = 0
+    for (let i = 0; i < memo.value.length; i++) {
+      count += isFullWidthChar(memo.value[i]) ? 2 : 1
+    }
+    return count
+  })
+
+  const isLimitReached = computed((): boolean => {
+    return messageCharCount.value > AUTHOR_MESSAGE_LIMIT
```
Comment (nwingt) [inline composables/useMessageCharCount.ts:22]:
```suggestion
    return messageCharCount.value > limit
```

Comment (AuroraHuang22) [reply composables/useMessageCharCount.ts]: [🎨 Rename params](https://github.com/likecoin/nft-book-press/pull/212/commits/b75094ed84ccfc054282d913fbd08c9bf2eab11d)

File: composables/useMessageCharCount.ts (hunk 4 above)
Comment (AuroraHuang22) [reply composables/useMessageCharCount.ts]: [🎨 Rename params](https://github.com/likecoin/nft-book-press/pull/212/commits/b75094ed84ccfc054282d913fbd08c9bf2eab11d)

**Review by nwingt — APPROVED** [review]

File: pages/nft-book-store/collection/send/[collectionId].vue [hunk 6]
```diff
@@ -160,6 +167,9 @@ import { useNftStore } from '~/stores/nft'
 import { useCollectionStore } from '~/stores/collection'
 import { parseImageURLFromMetadata } from '~/utils'
 import { getNFTs, getNFTOwner, signExecNFTSendAuthz, signSendNFTs } from '~/utils/cosmos'
+import { useMessageCharCount } from '~/composables/useMessageCharCount'
+
+const AUTHOR_MESSAGE_LIMIT = 98
```
Comment (williamchong) [inline pages/nft-book-store/collection/send/[collectionId].vue:172]: Should this be in `constant`?

Comment (AuroraHuang22) [reply pages/nft-book-store/collection/send/[collectionId].vue]: [♻️ Declare AUTHOR_MESSAGE_LIMIT as a constant](https://github.com/likecoin/nft-book-press/pull/212/commits/bb6e73edf3347bad3097cf1e3eda128e2ebd25e3)

**Review by williamchong — APPROVED** [review]

**Pull request merged**